crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,648 @@
+        let mut new = self;
+        let mut views = new.views.clone().unwrap_or_default();
+        views.lock_mut().push_cloned(v);
+        new.views = Some(views);
+        new
+    }
+}
+
+impl SlidingSync {
+    /// Generate a new SlidingSyncBuilder with the same inner settings and views
+    /// but without the current state
+    pub fn new_builder_copy(&self) -> SlidingSyncBuilder {
+        let mut builder = SlidingSyncBuilder::default()
+            .client(self.client.clone())
+            .views(Arc::new(futures_signals::signal_vec::MutableVec::new_with_values(
+                self.views
+                    .lock_ref()
+                    .to_vec()
+                    .iter()
+                    .map(|v| {
+                        v.new_builder().build().expect("builder worked before, builder works now")
+                    })
+                    .collect(),
+            )))
+            .subscriptions(Arc::new(futures_signals::signal_map::MutableBTreeMap::with_values(
+                self.subscriptions.lock_ref().to_owned(),
+            )))
+            .to_owned();
+
+        if let Some(ref h) = self.homeserver {
+            builder.homeserver(h.clone());
+        }
+        builder
+    }
+
+    /// Subscribe to a given room.
+    ///
+    /// Note: this does not cancel any pending request, so make sure to only
+    /// poll the stream after you've altered this. If you do that during, it
+    /// might take one round trip to take effect.
+    pub fn subscribe(
+        &self,
+        room_id: Box<RoomId>,
+        settings: Option<sliding_sync_events::RoomSubscription>,
+    ) {
+        self.subscriptions.lock_mut().insert_cloned(room_id, settings.unwrap_or_default());
+    }
+
+    /// Unsubscribe from a given room.
+    ///
+    /// Note: this does not cancel any pending request, so make sure to only
+    /// poll the stream after you've altered this. If you do that during, it
+    /// might take one round trip to take effect.
+    pub fn unsubscribe(&self, room_id: Box<RoomId>) {
+        if let Some(prev) = self.subscriptions.lock_mut().remove(&room_id) {
+            self.unsubscribe.lock_mut().push_cloned(room_id);
+        }
+    }
+
+    fn handle_response(
+        &self,
+        resp: sliding_sync_events::Response,
+        views: &[SlidingSyncView],
+    ) -> anyhow::Result<UpdateSummary> {
+        self.pos.replace(Some(resp.pos));
+        let mut updated_views = Vec::new();
+
+        if let Some(ops) = resp.ops {
+            let mut mapped_ops = Vec::new();
+            mapped_ops.resize_with(views.len(), Vec::new);
+
+            for (idx, ops) in ops
+                .iter()
+                .filter_map(|r| r.deserialize().ok())
+                .fold(mapped_ops, |mut mp, op| {
+                    let idx: u32 =
+                        op.list.try_into().expect("the list index is convertible into u32");
+                    mp[idx as usize].push(op);
```

Comment:
The index here could be out of bounds (if the server is buggy or outright malicious), leading to a panic, right?